pyswarms/utils/functions/grid_search.py

```diff
@@ -0,0 +1,102 @@
+of possible values, and their cartesian product is taken to produce a grid
+of all possible combinations. These combinations are then tested to produce
+a list of objective function scores. The default of the optimize method
+returns the hyperparameters that yield the minimum score, yet maximum score
+can also be assessed.
+"""
+
+import operator as op
+import itertools
+import numpy as np
+
+class GridSearch():
+    """
+    Exhaustive search of optimal performance on selected objective function
+    over all combinations of specified hyperparameter values.
+
+    Parameters:
+        * c1 : float
+            cognitive parameter
+        * c2 : float
+            social parameter
+        * w : float
+            inertia parameter
+        * k : int
+            number of neighbors to be considered. Must be a
+            positive integer less than `n_particles`
+        * p: int {1,2}
+            the Minkowski p-norm to use. 1 is the
+            sum-of-absolute values (or L1 distance) while 2 is
+            the Euclidean (or L2) distance.
+    >>> options = {'c1': [1, 2, 3],
+                   'c2': [1, 2, 3],
+                   'w' : [2, 3, 5],
+                   'k' : [5, 10, 15],
+                   'p' : 1}
+    >>> g = GridSearch(LocalBestPSO, n_particles=40, dimensions=20,
+                       options=options, objective_func=sphere_func, iters=10)
+    >>> best_options = g.optimize()
+    >>> best_options['c1']
+    1
+    >>> best_options['c2']
+    1
+    """
+
+    def __init__(self, optimizer, n_particles, dimensions, options,
+                 objective_func, iters, bounds=None, velocity_clamp=None):
+        """Initializes with dict of key, parameter value pairs."""
+        self._optimizer = optimizer
+        self._n_particles = n_particles
+        self._dims = dimensions
+        self._options = options
+        self._bounds = bounds
+        self._vclamp = velocity_clamp
+        self._objective_func = objective_func
+        self._iters = iters
+
+    def generate_grid(self):
+        """Generates the grid of all hyperparameter value combinations."""
+
+        #Extract keys and values from options dictionary
+        params = self._options.keys()
+        items = [x if type(x) == list \
+                 else [x] for x in list(zip(*self._options.items()))[1]]
+
+        #Create list of cartesian products of hyperparameter values from options
+        list_of_products = list(itertools.product(*items))
+
+        #Return list of dicts for all hyperparameter value combinations
+        return [dict(zip(*[params, list(x)])) for x in list_of_products]
+
+    def optimize(self, maximum=False):
+        best_options = None
+        params = self._options.keys()
+        def generate_score(options):
+            f = self._optimizer(self._n_particles, self._dims, options,
+                                self._bounds, self._vclamp)
+            return f.optimize(self._objective_func, self._iters)[0]
```

Comment:
I haven't checked if there are state-effects by adding a closure inside `optimize()`. Hmmm...will it change if we just convert `generate_report()` into a private function `_generate_report()` outside of `optimize()`? 

It's just a matter of personal style, it still depends on you if you keep this or not. I think closures are good as well. 👍

Reply:
I moved `generate_score` out, and I think this reads better and is conceptually more in line with how both `generate_grid` and `generate_score` behave in relationship to `optimize` (providing functionality for the report generation). 

I like your use of the word 'report', and it has me wondering if it could be used, or something like it, instead of 'optimize'. That might help make it clear that we are reporting performance of an optimizer in this method, yet that reporting is actually 'optimizing' anything, just running the optimizer with different parameters. Maybe `generate_report`?